File: MultilanguageSDK/Python/Samples/NYX650/ToFExposureTimeSetGet/ToFExposureTimeSetGet.py
from pickle import FALSE, TRUE
import sys
currentPath =  sys.path[0]
pos = currentPath.find('Samples')
libpath = currentPath[:pos]
sys.path.append(libpath) #absolutely path

from API.ScepterDS_api import *
import time

camera = ScepterTofCam()


camera_count = camera.scGetDeviceCount(3000)
print("Get device count:", camera_count)
if camera_count <= 0:
    print("scGetDeviceCount scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples.")
    exit()

device_info=ScDeviceInfo()

if camera_count > 0:
    ret,device_infolist=camera.scGetDeviceInfoList(camera_count)
    if ret==0:
        device_info = device_infolist[0]
    else:
        print(' failed:' + ret)  
        exit()  
else: 
    print("there are no camera found")
    exit()

if  ScConnectStatus.SC_CONNECTABLE.value != device_info.status:
	print("connect status:",device_info.status)  
	print("Call scOpenDeviceBySN with connect status :",ScConnectStatus.SC_CONNECTABLE.value)
	exit()
else:
    print("serialNumber: "+str(device_info.serialNumber))
    print("ip: "+str(device_info.ip))
    print("connectStatus: "+str(device_info.status))

ret = camera.scOpenDeviceBySN(device_info.serialNumber)
if  ret == 0:
    print("scOpenDeviceBySN")
else:
    print('scOpenDeviceBySN failed: ' + str(ret))   

ret = camera.scStartStream()
if  ret == 0:
    print("scStartStream successful")
else:
    print("scStartStream failed:"+ str(ret))   

'''Testing TOF exposure time requires turning off HDR and WDR in advance.'''
ret = camera.scSetExposureControlMode()
if  ret == 0:
    print("set tofsensor to manual mode successful")
else:
    print("scSetExposureControlMode failed:"+ str(ret))   

ret,frameRate = camera.scGetFrameRate()
if  ret == 0:
    print("Get default frame rate:"+ str(frameRate))   
else:
    print("scGetFrameRate failed:"+ str(ret))   

ret,MaxExposureTime = camera.scGetMaxExposureTime(ScSensorType.SC_TOF_SENSOR)
if  ret == 0:
    print("Recommended scope: 58 - "+ str(MaxExposureTime))   
else:
    print("scGetMaxExposureTime failed:"+ str(ret))   

ret = camera.scSetExposureTime(ScSensorType.SC_TOF_SENSOR,400)
if  ret == 0:
    print("Set exposure time 400 is ok")   
else:
    print("scSetExposureTime failed:"+ str(ret))   

ret = camera.scSetFrameRate(5)
if  ret == 0:
    print("Set frame rate 5 is ok")   
else:
    print("scSetFrameRate failed:"+ str(ret))   

ret,MaxExposureTime = camera.scGetMaxExposureTime(ScSensorType.SC_TOF_SENSOR)
if  ret == 0:
    print("Recommended scope: 58 - "+ str(MaxExposureTime))   
else:
    print("scGetMaxExposureTime failed:"+ str(ret))   

ret = camera.scSetExposureTime(ScSensorType.SC_TOF_SENSOR,500)
if  ret == 0:
    print("Set exposure time 500 is ok")   
else:
    print("scSetExposureTime failed:"+ str(ret))   
	
ret = camera.scStopStream()
if  ret == 0:
    print("stopstream success")
else:
    print("stopstream failed",ret)

ret = camera.scCloseDevice()     
if  ret == 0:
    print("scCloseDevice successful")
else:
    print('scCloseDevice failed: ' + str(ret)) 

print('Test end, please reboot camera to restore the default settings')  

File: BaseSDK/Ubuntu/Samples/Base/DS77/ToFExposureTimeSetGet/ToFExposureTimeSetGet.cpp
#include <thread>
#include <iostream>
#include "Scepter_api.h"

using namespace std;

int main()
{
	cout << "---ToFExposureTimeSetGet---" << endl;

	uint32_t deviceCount;
	ScDeviceInfo* pDeviceListInfo = NULL;
	ScDeviceHandle deviceHandle = 0;
	ScStatus status = SC_OTHERS;

	status = scInitialize();
	if (status == ScStatus::SC_OK)
	{
		cout << "[scInitialize] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scInitialize] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	status = scGetDeviceCount(&deviceCount, 3000);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetDeviceCount] success, ScStatus(" << status << "). The device count is " << deviceCount << endl;
	}
	else
	{
		cout << "[scGetDeviceCount] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}
	if (0 == deviceCount)
	{
		cout << "[scGetDeviceCount] scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples." << endl;
		return -1;
	}

	pDeviceListInfo = new ScDeviceInfo[deviceCount];
	status = scGetDeviceInfoList(deviceCount, pDeviceListInfo);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetDeviceInfoList] success, ScStatus(" << status << ").";
		if (SC_CONNECTABLE != pDeviceListInfo[0].status)
		{
			cout << " The first device [status]: " << pDeviceListInfo[0].status << " does not support connection." << endl;
			delete[] pDeviceListInfo;
			pDeviceListInfo = NULL;
			return -1;
		}
	}
	else
	{
		cout << "[scGetDeviceInfoList] fail, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
		return -1;
	}

	cout << " The first deviceInfo, <serialNumber>: " << pDeviceListInfo[0].serialNumber
		<< ", <ip>: " << pDeviceListInfo[0].ip << ", <status>: " << pDeviceListInfo[0].status << endl;

	status = scOpenDeviceBySN(pDeviceListInfo[0].serialNumber, &deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scOpenDeviceBySN] success, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
	}
	else
	{
		cout << "[scOpenDeviceBySN] fail, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
		return -1;
	}

	status = scStartStream(deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scStartStream] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scStartStream] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	//Testing TOF exposure time requires turning off HDR and WDR in advance.
	cout << endl << "---1. Default frame rate---" << endl;
	status = scSetExposureControlMode(deviceHandle, SC_TOF_SENSOR, SC_EXPOSURE_CONTROL_MODE_MANUAL);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scSetExposureControlMode] success, ScStatus(" << status << "). Set SC_EXPOSURE_CONTROL_MODE_MANUAL success." << endl;
	}
	else
	{
		cout << "[scSetExposureControlMode] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	int defaultframeRate = 10;
	status = scGetFrameRate(deviceHandle, &defaultframeRate);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetFrameRate] success, ScStatus(" << status << "). The device frame rate is " << defaultframeRate << endl;
	}
	else
	{
		cout << "[scGetFrameRate] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	int maxExposureTime = 0;
	status = scGetMaxExposureTime(deviceHandle, SC_TOF_SENSOR, &maxExposureTime);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetMaxExposureTime] success, ScStatus(" << status << "). Recommended scope: 58 - " << maxExposureTime << endl;
	}
	else
	{
		cout << "[scGetMaxExposureTime] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	int exposureTime = 400;
	status = scSetExposureTime(deviceHandle, SC_TOF_SENSOR, exposureTime);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scSetExposureTime] success, ScStatus(" << status << "). Set exposure time " << exposureTime << " is OK." << endl;
	}
	else
	{
		cout << "[scSetExposureTime] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	cout << endl << "---2. Set frame rate to 5---" << endl;
	int frameRate = 5;
	status = scSetFrameRate(deviceHandle, frameRate);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scSetFrameRate] success, ScStatus(" << status << "). Set frame rate " << frameRate << " is OK." << endl;
	}
	else
	{
		cout << "[scSetFrameRate] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	status = scGetMaxExposureTime(deviceHandle, SC_TOF_SENSOR, &maxExposureTime);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetMaxExposureTime] success, ScStatus(" << status << "). Recommended scope: 58 - " << maxExposureTime << endl;
	}
	else
	{
		cout << "[scGetMaxExposureTime] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	exposureTime = 500;
	status = scSetExposureTime(deviceHandle, SC_TOF_SENSOR, exposureTime);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scSetExposureTime] success, ScStatus(" << status << "). Set exposure time " << exposureTime << " is OK." << endl;
	}
	else
	{
		cout << "[scSetExposureTime] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}
	cout << endl;

	status = scStopStream(deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scStopStream] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scStopStream] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	status = scCloseDevice(&deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scCloseDevice] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scCloseDevice] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	status = scShutdown();
	if (status == ScStatus::SC_OK)
	{
		cout << "[scShutdown] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scShutdown] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}
	cout << "---Please reboot camera to restore the default settings---" << endl;
	cout << "---End---" << endl;

	return 0;
}
